### PR TITLE
feat(search): support the FT.TAGVALS command

### PR DESCRIPTION
**Pull request by jonathanc-n**
Added FT.TagVals for search + tests

### CODE REVIEW

File: src/search/index_manager.h [hunk 1]
```diff
@@ -272,6 +273,62 @@ struct IndexManager {
 
     return Status::OK();
   }
-};
 
+  Status FieldValues(engine::Context &ctx, std::string_view index_name, std::string_view input_tag,
```
Comment (PragmaTwice) [inline src/search/index_manager.h]: We should iterate through the index data, instead of redis keys.

Comment (jonathanc-n) [reply src/search/index_manager.h]: @PragmaTwice Put in the fixes! thanks for the feedback :)

File: src/search/index_manager.h [hunk 2]
```diff
@@ -272,6 +273,58 @@ struct IndexManager {
...
+
+    std::unordered_set<std::string> matching_values;
+    util::UniqueIterator index_iter(ctx, ctx.DefaultScanOptions(), ColumnFamilyID::Search);
+
+    auto index_key = SearchKey(ns, index_name, field_name);
+    std::string field_prefix;
+    index_key.PutNamespace(&field_prefix);
+    SearchKey::PutType(&field_prefix, SearchSubkeyType::FIELD);
+    index_key.PutIndex(&field_prefix);
+    PutSizedString(&field_prefix, field_name);
```
Comment (PragmaTwice) [inline src/search/index_manager.h:304]: You can use `ConstructTagFieldData({}, {})` here.

File: src/search/index_manager.h [hunk 3]
```diff
@@ -272,6 +273,58 @@ struct IndexManager {
 
     return Status::OK();
   }
-};
 
+  Status FieldValues(engine::Context &ctx, std::string_view index_name, std::string_view input_tag,
```
Comment (PragmaTwice) [inline src/search/index_manager.h]:
```suggestion
  Status FieldValues(engine::Context &ctx, std::string_view index_name, std::string_view tag_field_name,
```

File: src/search/index_manager.h [hunk 4]
```diff
@@ -272,6 +273,58 @@ struct IndexManager {
...
+      auto key = index_iter->key();
+
+      if (!key.starts_with(field_prefix)) {
+        break;
+      }
+
+      Slice key_slice = key;
+      key_slice.remove_prefix(field_prefix.size());
+
+      Slice field_value_slice;
```
Comment (PragmaTwice) [inline src/search/index_manager.h]:
```suggestion
      Slice tag;
```

File: src/search/index_manager.h [hunk 5]
```diff
@@ -272,6 +273,58 @@ struct IndexManager {
...
+      if (!GetSizedString(&key_slice, &field_value_slice)) continue;
+
+      matching_values.insert(field_value_slice.ToString());
```
Comment (PragmaTwice) [inline src/search/index_manager.h]: Here we can make an optimization: if the tag is still the last one, we can skip it instead of insert it.

Comment (jonathanc-n) [reply src/search/index_manager.h]: I have now implemented it, let me know if its fine

File: src/search/index_manager.h [hunk 6]
```diff
@@ -272,6 +273,67 @@ struct IndexManager {
 
     return Status::OK();
   }
-};
 
+  Status FieldValues(engine::Context &ctx, std::string_view index_name, std::string_view input_tag,
```
Comment (PragmaTwice) [inline src/search/index_manager.h]:
```suggestion
  StatusOr<std::unordered_set<std::string>> FieldValues(engine::Context &ctx, std::string_view index_name, std::string_view input_tag,
```

Comment (jonathanc-n) [reply src/search/index_manager.h]: I'm a little confused, how come this is the implementation here, however in other parts of Kvrocks, it uses a ret value? When do I know when to use Status + *ret value compared to using StatusOr<T>?

Comment (PragmaTwice) [reply src/search/index_manager.h]: It's preferred to use `StatusOr` instead of `Status` and pointer output.

File: src/commands/cmd_search.cc [hunk 7]
```diff
@@ -488,6 +504,7 @@ REDIS_REGISTER_COMMANDS(Search,
                         MakeCmdAttr<CommandFTExplain>("ft.explain", -3, "read-only", 0, 0, 0),
                         MakeCmdAttr<CommandFTInfo>("ft.info", 2, "read-only", 0, 0, 0),
                         MakeCmdAttr<CommandFTList>("ft._list", 1, "read-only", 0, 0, 0),
-                        MakeCmdAttr<CommandFTDrop>("ft.dropindex", 2, "write exclusive no-multi no-script", 0, 0, 0));
+                        MakeCmdAttr<CommandFTDrop>("ft.dropindex", 2, "write exclusive no-multi no-script", 0, 0, 0),
+                        MakeCmdAttr<CommandFTTagVals>("ft.tagvals", 3, "read-only", 0, 0, 0));
```
Comment (PragmaTwice) [inline src/commands/cmd_search.cc]:
```suggestion
                        MakeCmdAttr<CommandFTTagVals>("ft.tagvals", 3, "read-only slow", 0, 0, 0));
```

**Review by PragmaTwice — COMMENTED** [review]
The rest looks good to me.

File: src/search/index_manager.h [hunk 8]
```diff
@@ -272,6 +273,66 @@ struct IndexManager {
 
     return Status::OK();
   }
-};
 
+  StatusOr<std::unordered_set<std::string>> FieldValues(engine::Context &ctx, std::string_view index_name,
+                                                        std::string_view input_tag, const std::string &ns) {
```
Comment (PragmaTwice) [inline src/search/index_manager.h]:
```suggestion
                                                        std::string_view tag_field_name, const std::string &ns) {
```

**Comment by jonathanc-n** [comment]
@PragmaTwice Commited to change. Should be good to run workflow

**Comment by PragmaTwice** [comment]
Could you check the CI failure to see if it can be fixed?

**Comment by jonathanc-n** [comment]
@PragmaTwice CI, should be fixed. It was just incorrect return values for type mismatch.

**Review by PragmaTwice — APPROVED** [review]

**Pull request merged**